core/src/avm2/array.rs

```diff
@@ -33,49 +97,175 @@ impl<'gc> ArrayStorage<'gc> {
+                            storage.insert(item, value);
+                        }
+                    } else {
+                        storage.resize(item + 1, None);
+                        *storage.get_mut(item).unwrap() = Some(value);
+                        let new_holes = (item + 1) - storage.len();
+                        *occupied_count = (*occupied_count + new_holes) + 1;
+                    }
+                } else {
+                    if storage[item].is_none() {
+                        *occupied_count += 1;
+                    }
+                    *storage.get_mut(item).unwrap() = Some(value);
+                }
+            }
+            ArrayStorage::Sparse { storage, length } => {
+                storage.insert(item, value);
+                if item >= *length {
+                    *length = item + 1;
+                }
+            }
         }
-
-        *self.storage.get_mut(item).unwrap() = Some(value)
     }
 
     /// Delete an array storage slot, leaving a hole.
     pub fn delete(&mut self, item: usize) {
-        if let Some(i) = self.storage.get_mut(item) {
-            *i = None;
+        match self {
+            ArrayStorage::Dense {
+                storage,
+                occupied_count,
+            } => {
+                if let Some(i) = storage.get_mut(item) {
+                    *i = None;
+                    if *occupied_count != 0 {
+                        *occupied_count -= 1;
+                        self.maybe_convert_to_sparse();
+                    }
+                }
+            }
+            ArrayStorage::Sparse { storage, .. } => {
+                storage.remove(&item);
+                self.maybe_convert_to_dense();
+            }
         }
     }
 
     /// Get the length of the array.
     pub fn length(&self) -> usize {
-        self.storage.len()
+        match &self {
+            ArrayStorage::Dense { storage, .. } => storage.len(),
+            ArrayStorage::Sparse { length, .. } => *length,
+        }
     }
 
     /// Set the length of the array.
     pub fn set_length(&mut self, size: usize) {
-        self.storage.resize(size, None)
+        match self {
+            ArrayStorage::Dense {
+                storage,
+                occupied_count,
+            } => {
+                if Self::can_convert_to_sparse(size, *occupied_count) {
+                    self.convert_to_sparse();
+                    if let ArrayStorage::Sparse { .. } = self {
+                        self.set_length(size);
+                    }
+                } else {
+                    storage.resize(size, None);
+                }
+            }
+            ArrayStorage::Sparse { storage, length } => {
+                if size > *length {
+                    *length = size;
+                } else {
+                    let mut to_remove = Vec::new();
+                    for (i, _) in storage.range(size..) {
+                        to_remove.push(*i);
```

Comment:
I think this push-remove logic can be replaced by `storage.retain(|&k, _| k < size)`?